components/remote_settings/src/storage.rs

```diff
@@ -148,32 +148,83 @@ impl Storage {
+    pub fn merge_records(
+        &mut self,
+        collection_url: &str,
+        records: &Vec<RemoteSettingsRecord>,
+    ) -> Result<()> {
+        let tx = self.conn.transaction()?;
 
+        // Delete ALL existing records and metadata for with different collection_urls.
+        //
+        // This way, if a user (probably QA) switches the remote settings server in the middle of a
+        // browser sessions, we'll delete the stale data from the previous server.
+        tx.execute(
+            "DELETE FROM records where collection_url <> ?",
+            [collection_url],
+        )?;
+        tx.execute(
+            "DELETE FROM collection_metadata where collection_url <> ?",
+            [collection_url],
+        )?;
+        let max_last_modified = Self::update_record_rows(&tx, collection_url, records)?;
+        Self::update_collection_metadata(&tx, collection_url, max_last_modified)?;
+        tx.commit()?;
+        Ok(())
+    }
+
+    /// Insert/remove/update rows in the records table based on a records list
+    ///
+    /// Returns the max last modified record from the list
+    fn update_record_rows(
+        tx: &Transaction<'_>,
+        collection_url: &str,
+        records: &[RemoteSettingsRecord],
+    ) -> Result<u64> {
         // Find the max last_modified time while inserting records
         let mut max_last_modified = 0;
         {
-            let mut stmt =
-                tx.prepare("INSERT INTO records (id, collection_url, data) VALUES (?, ?, ?)")?;
+            let mut insert_stmt = tx.prepare(
+                "INSERT OR REPLACE INTO records (id, collection_url, data) VALUES (?, ?, ?)",
+            )?;
+            let mut delete_stmt = tx.prepare("DELETE FROM records WHERE id=?")?;
             for record in records {
-                max_last_modified = max_last_modified.max(record.last_modified);
-                let data = serde_json::to_vec(record)?;
-
-                stmt.execute(params![record.id, collection_url, data])?;
+                if record.deleted {
+                    delete_stmt.execute(params![&record.id])?;
+                } else {
+                    max_last_modified = max_last_modified.max(record.last_modified);
+                    let data = serde_json::to_vec(&record)?;
+                    insert_stmt.execute(params![record.id, collection_url, data])?;
+                }
             }
         }
+        Ok(max_last_modified)
+    }
 
+    /// Update the collection metadata after setting/merging records
+    fn update_collection_metadata(
+        tx: &Transaction<'_>,
+        collection_url: &str,
+        last_modified: u64,
+    ) -> Result<()> {
         // Update the metadata
         let fetched = true;
         tx.execute(
             "INSERT OR REPLACE INTO collection_metadata (collection_url, last_modified, fetched) VALUES (?, ?, ?)",
-            (collection_url, max_last_modified, fetched),
+            (collection_url, last_modified, fetched),
```

Comment:
@gruberb it seems like `fetched` is set to true when the row is inserted, do you think we could remove it and instead just check the presence of the row?  It's not really a big deal either way, the main reason I ask is that I want to start off some migrations for this component.

Reply:
I was also surprised with this `fetched` column. A count would do it